internal/dbtools/notification_preferences.go

```diff
@@ -0,0 +1,293 @@
+type UserNotificationPreference struct {
+	NotificationType string `json:"notification_type" boil:"notification_type"`
+	Enabled          *bool  `json:"enabled"`
+
+	NotificationTargets UserNotificationPreferenceTargets `json:"notification_targets"`
+}
+
+// UserNotificationPreferences is an alias for user notification
+// preference slice
+type UserNotificationPreferences []*UserNotificationPreference
+
+// RefreshNotificationDefaults refreshes the notification_defaults
+// materialized view
+func RefreshNotificationDefaults(ctx context.Context, ex boil.ContextExecutor) (err error) {
+	q := queries.Raw("REFRESH MATERIALIZED VIEW notification_defaults")
+	_, err = q.ExecContext(ctx, ex)
+
+	return
+}
+
+// GetNotificationPreferences fetch a user's notification preferences from
+// the governor DB, this function only gets rules the user had modified if
+// `withDefaults` is false
+func GetNotificationPreferences(ctx context.Context, uid string, ex boil.ContextExecutor, withDefaults bool) (UserNotificationPreferences, error) {
+	type preferencesQueryRecord struct {
+		NotificationType        string          `boil:"notification_type"`
+		NotificationTargetsJSON json.RawMessage `boil:"notification_targets"`
+	}
+
+	type preferencesQueryRecordNotificationTarget struct {
+		Target  string `json:"f1"`
+		Enabled *bool  `json:"f2"`
+	}
+
+	np := `
+		WITH np as (
+			SELECT 
+				user_id,
+				notification_types.id as type_id,
+				notification_types.slug as type_slug,
+				notification_preferences.notification_target_id_null_string as target_id,
+				notification_targets.slug as target_slug,
+				enabled
+			FROM notification_preferences
+			LEFT JOIN notification_targets ON notification_target_id = notification_targets.id
+			LEFT JOIN notification_types ON notification_type_id = notification_types.id
+			WHERE notification_preferences.user_id = $1
+		)
+	`
+
+	qWithDefaults := fmt.Sprintf(
+		"%s\n%s", np,
+		`
+		SELECT
+			nd.type_slug AS notification_type,
+			jsonb_agg((nd.target_slug, IFNULL(np.enabled, nd.default_enabled))) AS notification_targets
+		FROM notification_defaults as nd
+		FULL OUTER JOIN np on (np.target_id = nd.target_id AND np.type_id = nd.type_id)
+		GROUP BY nd.type_slug
+		`,
+	)
+
+	qWithoutDefaults := fmt.Sprintf(
+		"%s\n%s", np,
+		`
+		SELECT
+			np.type_slug AS notification_type,
+			jsonb_agg((np.target_slug, np.enabled)) AS notification_targets
+		FROM np
+		GROUP BY np.type_slug
+		`,
+	)
+
+	var q *queries.Query
+	if withDefaults {
+		q = queries.Raw(qWithDefaults, uid)
+	} else {
+		q = queries.Raw(qWithoutDefaults, uid)
+	}
+
+	records := []*preferencesQueryRecord{}
+
+	if err := q.Bind(ctx, ex, &records); err != nil {
+		return nil, err
+	}
+
+	eg := &errgroup.Group{}
+	preferences := make(UserNotificationPreferences, len(records))
+
+	for i, r := range records {
+		i, r := i, r // https://golang.org/doc/faq#closures_and_goroutines
+
+		eg.Go(func() error {
```

Comment:
did you explore how long this takes? looking at that package, it seems like negative limits remove the limt and a positive limit will ensure only x goroutines run at a time, but I'm not sure what the zero value limit is?